### PR TITLE
upstream: refactor logical/real host

### PR DESCRIPTION
**Pull request by mattklein123**
This change refactors logical/real host to be a common
concept that can be used by the dynamic forward proxy support.
In doing so, the threading has been switched from using TLS to
R/W locks to make dropping the code into multiple locations
easier. It also fixes a longstanding issue with logical DNS
where the current address is reported as "0.0.0.0" in admin
output.

Fixes https://github.com/envoyproxy/envoy/issues/5225
Part of https://github.com/envoyproxy/envoy/issues/1606

Signed-off-by: Matt Klein <mklein@lyft.com>

Risk Level: Medium
Testing: Existing and modified tests
Docs Changes: N/A
Release Notes: Added

### CODE REVIEW

File: source/common/upstream/logical_host.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+              const envoy::api::v2::endpoint::LocalityLbEndpoints& locality_lb_endpoint,
+              const envoy::api::v2::endpoint::LbEndpoint& lb_endpoint)
+      : HostImpl(cluster, hostname, address, lb_endpoint.metadata(),
+                 lb_endpoint.load_balancing_weight().value(), locality_lb_endpoint.locality(),
+                 lb_endpoint.endpoint().health_check_config(), locality_lb_endpoint.priority(),
+                 lb_endpoint.health_status()) {}
+
+  // Set the new address. Updates are typically rare so a R/W lock is used to address updates.
+  // Note that the health check address update requires no lock to be held since it is only
+  // used on the main thread.
```
Comment (snowp) [inline source/common/upstream/logical_host.h]: Do we have any kind of assert that ensures that this is true? Seems like a thing that could cause subtle issues if that ever changes. Or maybe we could rely on TSAN to catch that?

File: source/common/upstream/logical_host.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * A real host that forwards most of its calls to a logical host, but returns a snapped address.
+ */
+class RealHostDescription : public HostDescription {
+public:
+  RealHostDescription(Network::Address::InstanceConstSharedPtr address,
+                      HostConstSharedPtr logical_host)
+      : address_(address), logical_host_(logical_host) {}
+
+  // Upstream:HostDescription
+  bool canary() const override { return false; }
```
Comment (snowp) [inline source/common/upstream/logical_host.h]: I know this is existing code, but should this forward to the logical host?

File: source/common/upstream/logical_dns_cluster.cc [hunk 3]
```diff
@@ -87,20 +83,9 @@ void LogicalDnsCluster::startResolve() {
               Network::Utility::getAddressWithPort(*address_list.front(),
                                                    Network::Utility::portFromTcpUrl(dns_url_));
           if (!logical_host_) {
-            // TODO(mattklein123): The logical host is only used in /clusters admin output. We used
-            // to show the friendly DNS name in that output, but currently there is no way to
-            // express a DNS name inside of an Address::Instance. For now this is OK but we might
-            // want to do better again later.
-            switch (address_list.front()->ip()->version()) {
```
Comment (snowp) [inline source/common/upstream/logical_dns_cluster.cc]: How come we don't care about the ip version anymore?

File: source/common/upstream/logical_host.h (hunk 1 above)
Comment (mattklein123) [reply source/common/upstream/logical_host.h]: Let me just add locking to be on the safe side. I don't think it's worth it not to now that you mention it.

File: source/common/upstream/logical_host.h (hunk 2 above)
Comment (mattklein123) [reply source/common/upstream/logical_host.h]: Will do.

File: source/common/upstream/logical_dns_cluster.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/upstream/logical_dns_cluster.cc]: It will come directly from the DNS resolver so we don't need this 0.0.0.0 hack anymore.

**Pull request merged**